services/dis-apim-operator/internal/controller/api_controller.go

```diff
@@ -47,17 +62,257 @@
+	logger := log.FromContext(ctx)
+	var apiVersionList apimv1alpha1.ApiVersionList
+	apiVersionErr := r.List(ctx, &apiVersionList, client.InNamespace(api.Namespace), client.MatchingFields{jobOwnerKey: api.Name})
+	versions := apiVersionList.Items
+	if client.IgnoreNotFound(apiVersionErr) != nil {
+		logger.Error(apiVersionErr, "Failed to list versions")
+		return false
+	}
+	if len(versions) > 0 {
+		logger.Info("Versions still exist", "Count", len(versions))
+		return false
+	}
+	logger.Info("All versions deleted")
+	return true
+}
+
+func (r *ApiReconciler) handleCreateUpdate(ctx context.Context, api *apimv1alpha1.Api) (ctrl.Result, error) {
+	azApi, err := r.apimClient.CreateUpdateApiVersionSet(ctx, api.GetApiAzureFullName(), api.ToAzureApiVersionSet(), nil)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if azApi.ID == nil {
+		return ctrl.Result{}, fmt.Errorf("No ID returned for API version set %s", api.GetApiAzureFullName())
+	}
+	api.Status.ApiVersionSetID = *azApi.ID
+	api.Status.ProvisioningState = apimv1alpha1.ProvisioningStateUpdating
+	return ctrl.Result{RequeueAfter: 10 * time.Second}, r.Status().Update(ctx, api)
+}
+
+func (r *ApiReconciler) reconcileVersions(ctx context.Context, api *apimv1alpha1.Api) (bool, error) {
+	logger := log.FromContext(ctx)
+	wantedVersions := api.ToApiVersions()
+	inSync := true
+	if api.Status.VersionStates == nil && len(wantedVersions) > 0 {
+		api.Status.VersionStates = make(map[string]apimv1alpha1.ApiVersionStatus)
+	}
+	for k, wantedVersion := range wantedVersions {
+		var existingVersion apimv1alpha1.ApiVersion
+		if err := r.Get(ctx, client.ObjectKey{Namespace: api.Namespace, Name: wantedVersion.Name}, &existingVersion); err != nil {
+			if client.IgnoreNotFound(err) != nil {
+				logger.Error(err, "Failed to get ApiVersion")
+				return false, err
+			}
+			if err := controllerutil.SetControllerReference(api, &wantedVersion, r.Scheme); err != nil {
+				logger.Error(err, "Failed to set controller reference")
+				return false, err
+			}
+			if err := r.Create(ctx, &wantedVersion); err != nil {
+				logger.Error(err, "Failed to create ApiVersion")
+				return false, err
+			}
+			inSync = false
+			continue
+		} else {
+			if !wantedVersion.Matches(existingVersion) {
+				existingVersion.Spec = wantedVersion.Spec
+				if err := r.Update(ctx, &existingVersion); err != nil {
+					logger.Error(err, "Failed to update ApiVersion")
+					return false, err
+				}
+				inSync = false
+				api.Status.VersionStates[k] = existingVersion.Status
+			} else {
+				if existingVersion.Status.ProvisioningState != apimv1alpha1.ProvisioningStateSucceeded {
+					inSync = false
+				}
+				api.Status.VersionStates[k] = existingVersion.Status
+			}
+		}
+	}
+	if !inSync {
+		api.Status.ProvisioningState = getStatusFromVersionStatuses(api.Status.VersionStates)
+	} else {
+		api.Status.ProvisioningState = apimv1alpha1.ProvisioningStateSucceeded
+	}
+	return inSync, r.Status().Update(ctx, api)
+}
```

Comment:
I think this should be fine, one reconcile loop has one thread. If two reconciles happen at the same time this would operate on different objects and the last update would fail as it is trying to update an outdated resource in k8s